src/Components/Components/src/Routing/Router.cs

```diff
@@ -227,7 +228,24 @@ private async ValueTask<bool> RunOnNavigateAsync(string path, Task previousOnNav
             }
 
             var completedTask = await Task.WhenAny(task, cancellationTcs.Task);
-            return task == completedTask;
+
+            if (completedTask.IsFaulted)
+            {
+                ThrowExceptionInRenderer(completedTask);
+                return false;
+            }
+
+            return task == completedTask && !(completedTask.IsCanceled || _onNavigateCts.Token.IsCancellationRequested);
```

Comment:
This line is giving me quite a mental workout trying to understand the possible combinations 😄 

It's totally possible that the logic you have here is exactly correct already but at 11pm I'm lacking the mental agility to work that out. What would you think about writing this in a more step-by-step kind of way with comments explaining each of the sub-cases? For example, replacing lines 232-238 with this beast:

```cs
if (task == completedTask)
{
    try
    {
        await task;

        // A normal successful navigation
        return true;
    }
    catch (Exception e)
    {
        if (_onNavigateCts.Token.IsCancellationRequested)
        {
            // A normal superseded navigation in which 'task' happened to complete first, perhaps because the callback
            // passed the cancellation token through to something that auto-cancels
            return false;
        }
        else
        {
            // The failure/cancellation was not due to the navigation being superseded, so treat it as a fatal error
            // This applies regardless of whether the developer's task faulted or cancelled for a non-superseding-related reason
            ThrowExceptionInRenderer(completedTask);
            return false;
        }
    }
}
else
{
    // A normal superseded navigation in which cancellationTcs.Task happened to complete first
    return false;
}
```

BTW please don't trust this is exactly correct. I *think* it might well be but you'd be the one to know best.

Reply:
> This line is giving me quite a mental workout trying to understand the possible combinations 

I simplified this into a simpler statement in `task == completedTask && !completedTask.IsCompletedSuccessfully` which basically says `if the task that that finished first was our OnNavigateAsnyc and it wasn't in a cancelled state then return true.

I like that your code sample mimicks the pattern we use in the Async methods in ComponentBase but I'm not totally sure all the nesting is necessary and I like having the exception handling pulled out.

Reply:
Are we able to treat "`task` was cancelled but `cancellationTcs.Task` was not" as an error case (not just a no-op)?

Reply:
> Are we able to treat "task was cancelled but cancellationTcs.Task was not" as an error case (not just a no-op)?

Yes, we can. I believe we discussed this on the other PR but good to bring this feedback here.